File: 31_oct/problem_09_connected_components/solution.py
import sys
from typing import TextIO

class Solution(object):
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        # Build adjacency list
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        visited = [False] * n
        components = 0
        
        def dfs(node):
            visited[node] = True
            for neighbor in graph[node]:
                if not visited[neighbor]:
                    dfs(neighbor)
        
        for i in range(n):
            if not visited[i]:
                dfs(i)
                components += 1
        
        return components

def solve(infile=None, outfile=None):
    """Standalone execution for orchestrator"""
    if infile is None:
        infile = sys.stdin
    if outfile is None:
        outfile = sys.stdout
    
    n, m = map(int, infile.readline().split())
    edges = []
    for _ in range(m):
        line = infile.readline().strip()
        if line:  # Only process non-empty lines
            u, v = map(int, line.split())
            edges.append([u, v])
    
    solution = Solution()
    result = solution.countComponents(n, edges)
    outfile.write(str(result) + '\n')

if __name__ == "__main__":
    solve()

File: 31_oct/problem_09_connected_components/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <sstream>
using namespace std;

class Solution {
private:
    void dfs(int node, vector<vector<int>>& graph, vector<bool>& visited) {
        visited[node] = true;
        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor, graph, visited);
            }
        }
    }
    
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        // Build adjacency list
        vector<vector<int>> graph(n);
        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        
        vector<bool> visited(n, false);
        int components = 0;
        
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i, graph, visited);
                components++;
            }
        }
        
        return components;
    }
};

int main() {
    int n, m;
    cin >> n >> m;
    cin.ignore(); // Ignore the newline after n and m
    
    vector<vector<int>> edges;
    for (int i = 0; i < m; i++) {
        string line;
        getline(cin, line);
        if (!line.empty()) {
            istringstream iss(line);
            int u, v;
            iss >> u >> v;
            edges.push_back({u, v});
        }
    }
    
    Solution solution;
    int result = solution.countComponents(n, edges);
    cout << result << endl;
    
    return 0;
}
